<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
# DBNAME = 
# USER = 
# PASSWORD = 
# HOST = 
# PORT = 

DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select count(*)
            from public.VACANCIES 
         '''

In [5]:
# результат запроса
num_vacancies = pd.read_sql_query(query_3_1, connection)
print(num_vacancies)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\1716733680.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  num_vacancies = pd.read_sql_query(query_3_1, connection)


   count
0  49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select count(*)
            from public.employers 
         '''

In [7]:
# результат запроса
num_employers  = pd.read_sql_query(query_3_2, connection)
print(num_employers)

   count
0  23501


C:\Users\User\AppData\Local\Temp\ipykernel_8780\1238590485.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  num_employers  = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса

query_3_3 = f'''select count(*)
            from public.areas
         '''

In [9]:
# результат запроса
num_areas  = pd.read_sql_query(query_3_3, connection)
print(num_areas)

   count
0   1362


C:\Users\User\AppData\Local\Temp\ipykernel_8780\976750379.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  num_areas  = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select count(*)
            from public.industries
         '''

In [11]:
# результат запроса
num_industries  = pd.read_sql_query(query_3_4, connection)
print(num_industries)

   count
0    294


C:\Users\User\AppData\Local\Temp\ipykernel_8780\3301278474.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  num_industries  = pd.read_sql_query(query_3_4, connection)


***

In [12]:
# выводы по предварительному анализу данных

<span style="color:blue">
Для анализа нам предоставлена выборка из чуть более 49 тыс вакансий, а также справочники работодателей, регионов и отраслей.
<span>

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''select a.name as area, count(*) as cnt
            from public.vacancies v
            join public.areas a on v.area_id = a.id
            group by a.id
            order by cnt desc
            limit 5
         '''

In [14]:
# результат запроса
top_areas  = pd.read_sql_query(query_4_1, connection)
print(top_areas)

              area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892


C:\Users\User\AppData\Local\Temp\ipykernel_8780\557539298.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_areas  = pd.read_sql_query(query_4_1, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''select count(*)
            from public.vacancies v
            where salary_from is not null or
            salary_to is not null
         '''

In [16]:
# результат запроса
not_empty_salary  = pd.read_sql_query(query_4_2, connection)
print(not_empty_salary)

   count
0  24073


C:\Users\User\AppData\Local\Temp\ipykernel_8780\3580064510.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  not_empty_salary  = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''select avg(salary_from)::int as minVal, avg(salary_to)::int as maxVal
            from public.vacancies v
         '''

In [18]:
# результат запроса
avg_minmax_salary  = pd.read_sql_query(query_4_3, connection)
print(avg_minmax_salary)

   minval  maxval
0   71065  110537


C:\Users\User\AppData\Local\Temp\ipykernel_8780\186085068.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_minmax_salary  = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = f'''select schedule, employment, count(*) as cnt
            from public.vacancies 
            group by schedule, employment
            order by cnt desc
         '''

In [20]:
# результат запроса
schedule_employement  = pd.read_sql_query(query_4_4, connection)
print(schedule_employement)

            schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


C:\Users\User\AppData\Local\Temp\ipykernel_8780\444208574.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schedule_employement  = pd.read_sql_query(query_4_4, connection)


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = f'''select experience, count(*) as cnt
            from public.vacancies 
            group by experience
            order by cnt asc
         '''

In [22]:
# результат запроса
experience  = pd.read_sql_query(query_4_5, connection)
print(experience)

           experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152


C:\Users\User\AppData\Local\Temp\ipykernel_8780\1337973376.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience  = pd.read_sql_query(query_4_5, connection)


***

In [23]:
# выводы по детальному анализу вакансий

 <span style="color:blue">
 Крупные города существеннее представлены в данной выборке. Более 10% вакансий приходится на Москву, ~ 5,8% - Санкт-Петербург, ~ 4,3% всех вакансий - Минск.<br>
 Из всей нашей выборки только у 49% вакансий заполнена хотя бы одна из границ "зарплатной вилки". Для тех вакансий, где эти данные заполнены, мы можем увидеть, что средние границы зарплатных ожиданий - от 71065 до 110537 рублей в месяц. <br>
 Более 70% всех вакансий предплагают полную занятость и присутствие на рабочем месте полный рабочий день. Только 18,9% вакансий предоставляют возможность удаленной работы. <br>
 Больше половины всех вакансий предполагают наличие у соискателя хотя бы минимального трудового опыта (от 1 до 3 лет).
 </span>

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1 = f'''select e.name, count(*) as cnt
            from public.vacancies v
            join public.employers e on v.employer_id = e.id
            group by e.id
            order by cnt desc
            limit 5
         '''

In [25]:
# результат запроса
employers  = pd.read_sql_query(query_5_1, connection)
print(employers)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\3475601397.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers  = pd.read_sql_query(query_5_1, connection)


            name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = f'''select a.name, count(distinct e.id) as employers_cnt, count(distinct v.id) as vacancies_count
            from public.areas a
            left join public.employers e on a.id = e.area
            left join public.vacancies v on a.id = v.area_id
            group by a.id
            having count(distinct v.id) = 0 
            order by count(e.id) desc
            limit 1

         '''

In [27]:
# результат запроса
stats_by_area  = pd.read_sql_query(query_5_2, connection)
print(stats_by_area)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\777429006.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  stats_by_area  = pd.read_sql_query(query_5_2, connection)


     name  employers_cnt  vacancies_count
0  Россия            410                0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3 = f'''select e.name, count(distinct v.area_id) as area_cnt
            from public.employers e 
            left join public.vacancies v on v.employer_id = e.id
            group by e.id
            order by area_cnt desc
            limit 5

         '''

In [29]:
# результат запроса
employers_by_area  = pd.read_sql_query(query_5_3, connection)
print(employers_by_area)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\1013440459.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_by_area  = pd.read_sql_query(query_5_3, connection)


                     name  area_cnt
0                  Яндекс       181
1              Ростелеком       152
2              Спецремонт       116
3  Поляков Денис Иванович        88
4               ООО ЕФИН         71


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4 = f'''select count(*)
            from public.employers e
            left join public.employers_industries ei on ei.employer_id = e.id
            where ei.industry_id is null
         '''

In [31]:
# результат запроса
employers_no_indust  = pd.read_sql_query(query_5_4, connection)
print(employers_no_indust)

   count
0   8419


C:\Users\User\AppData\Local\Temp\ipykernel_8780\1510140253.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_no_indust  = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# # текст запроса
query_5_5= f'''select e.name
            from public.employers_industries ei
            join public.employers e on e.id = ei.employer_id
            group by e.id
            having count(ei.industry_id) = 4
            order by e.name
            limit 1 offset 2
        '''

In [33]:
# результат запроса
employers_4_indust  = pd.read_sql_query(query_5_5, connection)
print(employers_4_indust)

   name
0  2ГИС


C:\Users\User\AppData\Local\Temp\ipykernel_8780\3616763459.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_4_indust  = pd.read_sql_query(query_5_5, connection)


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6= f'''select count(*)
            from public.employers_industries ei
            join public.employers e on e.id = ei.employer_id
            join public.industries i on i.id = ei.industry_id
            where i.name = 'Разработка программного обеспечения'
        '''

In [35]:
# результат запроса
employers_software  = pd.read_sql_query(query_5_6, connection)
print(employers_software)

   count
0   3553


C:\Users\User\AppData\Local\Temp\ipykernel_8780\573261659.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_software  = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников

cities_list = list()

# Импортируем библиотеку BeautifulSoup
from bs4 import BeautifulSoup

# Импортируем библиотеку requests
import requests

# Определяем значение переменной url, содержащей адрес страницы для запроса
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'

# Выполняем GET-запрос, содержимое ответа присваивается переменной response
response = requests.get(url)

# Создаём объект BeautifulSoup, указывая html-парсер 
page = BeautifulSoup(response.text, 'html.parser')

# Получаем тег table
table = page.find('table', class_="standard sortable")

# Ищем все ячейки таблицы, а в них ссылки с названиями городов

for line in table.findAll('tr'):
    for cell in line.findAll('td'):
        for link in cell.findAll('a'):
            city = link.get('title')
            if city is not None:
                cities_list.append(city)

                       
city_string = str(cities_list)[1:-1]
print(city_string)

#cities_list = ['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']

'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград'


In [37]:
# текст запроса

# Отдельнымы запросами посчитаем кол-во вакансий с группировкой по городам и отдельные общее значение (Total). Результаты двух запросов объединим через union all и отсортируем.

query_5_7 = f'''(select a.name, count(*) as cnt
            from public.employers e 
            left join public.vacancies v on v.employer_id = e.id
            left join public.areas a on v.area_id = a.id
            where e.name ='Яндекс'
            and a.name in ({city_string})
            group by e.id, a.id)
            union all
            (select 'Total', count(*) as cnt
            from public.employers e 
            left join public.vacancies v on v.employer_id = e.id
            left join public.areas a on v.area_id = a.id
            where e.name ='Яндекс'
            and a.name in ({city_string})
            group by e.id)
            order by cnt asc
        '''
        

In [38]:
# результат запроса
yandex = pd.read_sql_query(query_5_7, connection)
print(yandex)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\2632274865.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex = pd.read_sql_query(query_5_7, connection)


               name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485


***

In [39]:
# выводы по анализу работодателей

<span style="color:blue">
В выборке представлено несколько крпных работодателей, которые имеют большое кол-во открытых вакансий (в лидерах - Яндекс с 3,9% от всех вакансий). Такие компании имеют широкую географию и представлены сразу в нескольких регионах. Иногда эта география столь обширна, что вакансии могут быть представлены более, чем в 150 населенных пунктах.<br>
У 17% работодателей не заполнена сфера их деятельности.  Но из тех, кто ее указал, 8,7% нашей выборки занимаются (в том числе) разработкой программного обеспечения.<br>
<br>
Если рассматривать распределение вакансий компании "Яндекс" по городам-миллионникам, то в наиболее крупных (столичных) городах вакансий представлено чуть больше, но тем не менее видно, что компания так же стремится набирать людей и более удаленных от столиц регионах с крупными административными центрами.
</span>

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''select count(*)
            from public.vacancies v
            where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
        '''

In [41]:
# результат запроса
data_vac = pd.read_sql_query(query_6_1, connection)
print(data_vac)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\3537974588.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vac = pd.read_sql_query(query_6_1, connection)


   count
0   1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
# текст запроса
query_6_2 = f'''select count(*)
            from public.vacancies v
            where ( lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and
                    (lower(v.name) like '%junior%'
                    or v.experience = 'Нет опыта'
                    or v.employment = 'Стажировка') 
        '''

In [43]:
# результат запроса
junior_data_scientist = pd.read_sql_query(query_6_2, connection)
print(junior_data_scientist)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\1070903414.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_data_scientist = pd.read_sql_query(query_6_2, connection)


   count
0     51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# текст запроса

query_6_3 = f''' select count(*) from public.vacancies v
                where (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgr%') and
                    (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%'  and lower(v.name) like '%teaml%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    
        '''
        

In [45]:
# результат запроса
sql_skills = pd.read_sql_query(query_6_3, connection)

#sql_skills.to_csv(r'D:\python/vacSql.txt', sep=';')
          
print(sql_skills)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\2370291304.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_skills = pd.read_sql_query(query_6_3, connection)


   count
0    201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [46]:
# текст запроса
query_6_4 = f'''select count(*)
            from public.vacancies v
            where ( lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and lower(v.key_skills) like '%python%' 
                    
        '''
        

In [47]:
# результат запроса
py_skills = pd.read_sql_query(query_6_4, connection)
print(py_skills)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\3534483240.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  py_skills = pd.read_sql_query(query_6_4, connection)


   count
0    351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [48]:
# текст запроса

# Посчитаем количество табуляций между ключевыми навыками и прибавим один (крайний).

query_6_5  = f'''select avg(t.cnt) from 
            (select name, v.key_skills, length(v.key_skills) - length(replace(v.key_skills, E'\t', '')) +1  as cnt
            from public.vacancies v
            where  (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and v.key_skills is not null and v.key_skills <> '') t
                
        '''
       

In [49]:
# результат запроса
avg_skills = pd.read_sql_query(query_6_5, connection)
print(avg_skills)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\3461183951.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills = pd.read_sql_query(query_6_5, connection)


        avg
0  6.417062


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [50]:
# текст запроса

# Объединим результаты выполнения трех подзапросов (где есть обе границы "зп вилки", где есть только верхняя и где есть только нижняя), найдя среднее значение для кажого из них.

query_6_6  = f'''select t.experience, avg(t.cnt) from

                (select  v.experience, (v.salary_from + v.salary_to)/2 as cnt
                from public.vacancies v
                 where  (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and salary_from is not null and salary_to is not null
                               
                union all
                
                select v.experience, v.salary_to
                from public.vacancies v
                where  (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and salary_from is null and salary_to is not null 
               
                
                union all
                
                select v.experience, v.salary_from
                from public.vacancies v
                where  (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and salary_from is not null and salary_to is null
                
               ) t
               group by t.experience
        '''

In [51]:
# результат запроса
avg_salary = pd.read_sql_query(query_6_6, connection)
print(avg_salary)

C:\Users\User\AppData\Local\Temp\ipykernel_8780\624173144.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_6_6, connection)


           experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  141103.321429
2       От 3 до 6 лет  243114.666667


***

In [52]:
# выводы по предметному анализу


<span style="color:blue">
3,6% всех вакансий в выборке так или иначе относятся к работе с данным. Из них 472 это вакансии для дата-сайентиста, из которых 51 для специалистов начального уровня.<br>
Для вакансий дата-сайентиста важными навыками являются знание python - 351 и SQL/базы данных - 201 вакансии соответственно.<br> В среднем же для специалиста по data science в вакансиях указывают порядка 6 ключевых навыков.<br>
Средняя зарплата специалиста по data science зависит от опыта соискателя и составляет от 74,6 тыс. рублей для начинающего специалиста и до 240+ тыс. рублей для опытного соискателя.
</span>


# Общий вывод по проекту

In [53]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

<span style="color:blue">
Можно продолжить исследования рынка вакансий по специализации Data Science так же проанализировав следующие аспекты:<br>
- в каких отраслях востребованы DS специалисты<br>
- есть ли отличия в требуемых ключевых навыках специалиста в зависимости от отрасли компании<br>
- какие условия труда предлагаются специалисту (есть ли возможность удаленной работы и неполного рабочего дня)<br>
- как меняется средняя зп специалиста в зависимости от региона<br>
и т.п.<br>
</span>

In [ ]:
query_add_1  = f'''select avg(t.cnt) from 
            from public.vacancies v
            where  (lower(v.name) like '%data%scientist%' 
                    or lower(v.name) like '%data%science%'  
                    or lower(v.name) like '%исследовател%данных%'  
                    or (lower(v.name) like '%ml%' and lower(v.name) not like '%html%' and lower(v.name) not like '%qml%' and lower(v.name) not like '%teamlead%')
                    or lower(v.name) like '%machine%learning%'  
                    or lower(v.name) like '%машинн%обучен%'
                    )
                and v.key_skills is not null and v.key_skills <> '') t
                
        '''